File: crates/networker-dashboard/src/services/vm_lifecycle_recorder.rs
//! Lib-side recorder for `vm_lifecycle` events.
//!
//! Lives in `services/` rather than `db/` because auto-shutdown and error
//! paths run from inside service modules (`tester_scheduler`,
//! `tester_state`), which are in the lib crate and can't see the bin-side
//! `db::vm_lifecycle` module due to the dashboard's lib/bin split.
//!
//! Uses raw SQL so it carries no struct dependencies on bin-side types
//! (`ProjectTesterRow` et al). Callers pass primitives directly — all
//! snapshot columns are string slices so we don't have to mirror row types
//! here either.
//!
//! `insert_event` mirrors the public contract of the bin-side
//! `db::vm_lifecycle::insert` helper: failures log at WARN and are
//! swallowed, since a transient audit-table issue must not surface as a
//! user-visible failure on the state change that triggered it.
//!
//! Event-type + resource-type strings here MUST stay in sync with the DB
//! CHECK constraints (`vm_lifecycle_event_type_valid`,
//! `vm_lifecycle_resource_type_valid`) added in migration V034, and with
//! the bin-side `db::vm_lifecycle::EventType` / `ResourceType` enums.

use chrono::{DateTime, Utc};
use tokio_postgres::Client;
use uuid::Uuid;

/// Minimal snapshot inputs for a lifecycle insert.
///
/// Intentionally primitive-typed so lib-side callers need no bin types to
/// use it. Matches the columns written by `db::vm_lifecycle::insert` so
/// rows produced by either path are indistinguishable on read.
#[derive(Debug, Clone)]
#[allow(clippy::too_many_arguments)]
pub struct TesterEventInput<'a> {
    pub project_id: &'a str,
    pub tester_id: Uuid,
    pub tester_name: &'a str,
    pub cloud: &'a str,
    pub region: &'a str,
    pub vm_size: &'a str,
    pub vm_name: Option<&'a str>,
    pub vm_resource_id: Option<&'a str>,
    pub cloud_connection_id: Option<Uuid>,
    pub event_type: &'a str,
    pub event_time: DateTime<Utc>,
    pub triggered_by: Option<Uuid>,
    pub metadata: Option<serde_json::Value>,
}

/// Append a `vm_lifecycle` row. `event_type` must be one of
/// `created | started | stopped | deleted | auto_shutdown | error` — the
/// DB CHECK constraint will reject anything else.
///
/// Never returns an error from the caller's perspective; logs at WARN on
/// failure so a transient DB issue can't crash the scheduler loop.
pub async fn insert_tester_event<'a>(client: &Client, event: TesterEventInput<'a>) {
    let result = client
        .execute(
            "INSERT INTO vm_lifecycle (
                project_id, resource_type, resource_id, resource_name,
                cloud, region, vm_size, vm_name, vm_resource_id,
                cloud_connection_id, event_type, event_time, triggered_by, metadata
            ) VALUES (
                $1, 'tester', $2, $3,
                $4, $5, $6, $7, $8,
                $9, $10, $11, $12, $13
            )",
            &[
                &event.project_id,
                &event.tester_id,
                &event.tester_name,
                &event.cloud,
                &event.region,
                &event.vm_size,
                &event.vm_name,
                &event.vm_resource_id,
                &event.cloud_connection_id,
                &event.event_type,
                &event.event_time,
                &event.triggered_by,
                &event.metadata,
            ],
        )
        .await;
    if let Err(e) = result {
        tracing::warn!(
            tester_id = %event.tester_id,
            event_type = %event.event_type,
            error = %e,
            "failed to append vm_lifecycle event from lib-side path (history incomplete, user-facing op unaffected)"
        );
    }
}

File: src/Networker.ControlPlane/Provisioning/VmLifecycleRecorder.cs
using Microsoft.EntityFrameworkCore;
using Networker.Data;
using Networker.Data.Entities;

namespace Networker.ControlPlane.Provisioning;

/// <summary>
/// Input for a tester lifecycle audit event — the C# port of Rust
/// <c>TesterEventInput</c> in
/// <c>crates/networker-dashboard/src/services/vm_lifecycle_recorder.rs</c>.
/// </summary>
/// <param name="ProjectId">Owning project id.</param>
/// <param name="TesterId">Tester id (becomes <c>resource_id</c>).</param>
/// <param name="TesterName">Tester name (becomes <c>resource_name</c>).</param>
/// <param name="Cloud">Cloud provider string.</param>
/// <param name="Region">Cloud region.</param>
/// <param name="VmSize">VM size / SKU.</param>
/// <param name="VmName">Cloud VM name, if known.</param>
/// <param name="VmResourceId">Cloud resource id, if known.</param>
/// <param name="CloudConnectionId">Owning cloud connection, if any.</param>
/// <param name="EventType">One of created|started|stopped|deleted|auto_shutdown|error
/// (DB CHECK constraint <c>vm_lifecycle_event_type_valid</c>).</param>
/// <param name="EventTime">When the event occurred (UTC).</param>
/// <param name="TriggeredBy">User who triggered it, if any.</param>
/// <param name="Metadata">Optional JSON metadata (serialized string).</param>
public sealed record TesterEventInput(
    string ProjectId,
    Guid TesterId,
    string TesterName,
    string Cloud,
    string Region,
    string VmSize,
    string? VmName,
    string? VmResourceId,
    Guid? CloudConnectionId,
    string EventType,
    DateTime EventTime,
    Guid? TriggeredBy,
    string? Metadata);

/// <summary>
/// Injectable audit recorder that appends <c>vm_lifecycle</c> rows on tester
/// state changes — the C# port of Rust
/// <c>crates/networker-dashboard/src/services/vm_lifecycle_recorder.rs</c>
/// (<c>insert_tester_event</c>).
///
/// <para><b>Faithful behavior:</b> failures are <b>swallowed</b> and logged at
/// WARN — never surfaced to the caller (audit history may be incomplete, but the
/// user-facing op is unaffected). <c>resource_type</c> is hardcoded to
/// <c>'tester'</c>; <c>resource_id ← tester_id</c>, <c>resource_name ← tester_name</c>,
/// exactly as the Rust INSERT.</para>
///
/// <para><b>Call sites it SHOULD be wired into</b> (document only — those files
/// are NOT edited in this port): the tester lifecycle write path
/// (<c>TesterWriteEndpoints</c>) — on each tester power/allocation change, emit
/// the matching event: <c>created</c> (provision), <c>started</c> (start),
/// <c>stopped</c> (stop), <c>deleted</c> (delete), <c>auto_shutdown</c>
/// (auto-shutdown loop), <c>error</c> (transition to error). Inject
/// <see cref="IVmLifecycleRecorder"/> and call
/// <see cref="RecordTesterEventAsync"/> after the state write commits.</para>
/// </summary>
public interface IVmLifecycleRecorder
{
    /// <summary>Append a tester lifecycle audit row. Best-effort — never throws.</summary>
    Task RecordTesterEventAsync(TesterEventInput input, CancellationToken ct = default);
}

/// <inheritdoc cref="IVmLifecycleRecorder"/>
public sealed class VmLifecycleRecorder : IVmLifecycleRecorder
{
    private readonly NetworkerDbContext _db;
    private readonly ILogger<VmLifecycleRecorder> _logger;

    public VmLifecycleRecorder(NetworkerDbContext db, ILogger<VmLifecycleRecorder> logger)
    {
        _db = db;
        _logger = logger;
    }

    public async Task RecordTesterEventAsync(TesterEventInput input, CancellationToken ct = default)
    {
        try
        {
            var row = new VmLifecycle
            {
                EventId = Guid.NewGuid(),
                ProjectId = input.ProjectId,
                ResourceType = "tester", // hardcoded literal, matching Rust
                ResourceId = input.TesterId,
                ResourceName = input.TesterName,
                Cloud = input.Cloud,
                Region = input.Region,
                VmSize = input.VmSize,
                VmName = input.VmName,
                VmResourceId = input.VmResourceId,
                CloudConnectionId = input.CloudConnectionId,
                EventType = input.EventType,
                EventTime = input.EventTime,
                TriggeredBy = input.TriggeredBy,
                Metadata = input.Metadata,
                CreatedAt = DateTime.UtcNow,
            };

            _db.VmLifecycles.Add(row);
            await _db.SaveChangesAsync(ct).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            // Rust: WARN "failed to append vm_lifecycle event from lib-side path
            // (history incomplete, user-facing op unaffected)".
            _logger.LogWarning(ex,
                "failed to append vm_lifecycle event (history incomplete, user-facing op unaffected) " +
                "tester_id={TesterId} event_type={EventType}",
                input.TesterId, input.EventType);
        }
    }
}

/// <summary>
/// DI wiring for <see cref="IVmLifecycleRecorder"/>. Scoped (it holds the scoped
/// <see cref="NetworkerDbContext"/>). Add in <c>Program.cs</c>:
/// <code>builder.Services.AddVmLifecycleRecorder();</code>
/// </summary>
public static class VmLifecycleRecorderExtensions
{
    public static IServiceCollection AddVmLifecycleRecorder(this IServiceCollection services)
    {
        services.AddScoped<IVmLifecycleRecorder, VmLifecycleRecorder>();
        return services;
    }
}
